src/Build/Evaluation/ProjectRootElementCache.cs

```diff
@@ -460,7 +460,6 @@ internal override void DiscardImplicitReferences()
                         }
                         else
                         {
-                            _strongCache.Remove(rootElement);
                             RaiseProjectRootElementRemovedFromStrongCache(rootElement);
```

Comment:
[nit / pre-existing issue] What guarantees that this PRE was in the old strong cache and is actually being removed? It looks like we also get here if it was a regular weak-only not-explicitly-loaded entry.

Reply:
Good point. There are multiple ways we could go down this path and erroneously call the remove event. I wanted to make this an unambiguous no-op, so I stopped short of removing the next call, which might be legitimate. If I were to rework it to be more correct, I think this is best:

```C#
if (rootElement.IsExplicitlyLoaded)
{
    if (oldStrongCache.Contains(rootElement))
    {
        _strongCache.AddFirst(rootElement);
    }
    else
    {
        // Has already been removed. We can consider adding it to the strong cache here...but it probably isn't worth it
    }
}
else
{
    if (oldStrongCache.Contains(rootElement))
    {
        RaiseProjectRootElementRemovedFromStrongCache(rootElement);
    }
    else
    {
        // Element was not in the strong cache and should not be added to it --> no op
    }
}
```

or more simply

```C#
if (oldStrongCache.Contains(rootElement))
{
    if (rootElement.IsExplicitlyLoaded)
    {
        _strongCache.AddFirst(rootElement);
    }
    else
    {
        RaiseProjectRootElementRemovedFromStrongCache(rootElement);
    }
}
```

I think either is better from a correctness perspective, but it means we'd always be calling Contains, which is a bit slow.

Reply:
What if we make it iterate over the strong cache? Should be doable in one pass. And as a bonus, it would preserve the order, which the current implementation does not.

Reply:
The only issue I see with that is that we're also discarding implicit references in the weak cache in this pass, and since there can be ProjectRootElements in the weak cache but not the strong cache, we'd be deleting those from the weak cache whether they're explicitly loaded and not GC'd or not.